u_parrots.go

```diff
@@ -938,3 +941,367 @@ func removeRC4Ciphers(s []uint16) []uint16 {
+				clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &GenericExtension{extension, sigAndAlgs})
+				// end utls specifis
+
+			case extensionRenegotiationInfo:
+				// RFC 5746, Section 3.2
+				if !readUint8LengthPrefixed(&extData, &m.secureRenegotiation) {
+					return false
+				}
+				m.secureRenegotiationSupported = true
+
+				// utls specific
+				clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &RenegotiationInfoExtension{RenegotiateFreelyAsClient})
+				// end utls specific
+
+			case extensionALPN:
+				// RFC 7301, Section 3.1
+				var protoList cryptobyte.String
+				if !extData.ReadUint16LengthPrefixed(&protoList) || protoList.Empty() {
+					return false
+				}
+				for !protoList.Empty() {
+					var proto cryptobyte.String
+					if !protoList.ReadUint8LengthPrefixed(&proto) || proto.Empty() {
+						return false
+					}
+					m.alpnProtocols = append(m.alpnProtocols, string(proto))
+
+				}
+
+				// utls specific
+				clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &ALPNExtension{m.alpnProtocols})
+				// end utls specific
+
+			case extensionSCT:
+				// RFC 6962, Section 3.3.1
+				m.scts = true
+
+				// utls specific
+				clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &SCTExtension{})
+				// end utls specific
+
+			case extensionSupportedVersions:
+				// RFC 8446, Section 4.2.1
+				var versList cryptobyte.String
+				if !extData.ReadUint8LengthPrefixed(&versList) || versList.Empty() {
+					return false
+				}
+				for !versList.Empty() {
+					var vers uint16
+					if !versList.ReadUint16(&vers) {
+						return false
+					}
+					m.supportedVersions = append(m.supportedVersions, vers)
+				}
+
+				// utls specific
+				clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &SupportedVersionsExtension{m.supportedVersions})
+				// end utls specific
+
+			case extensionCookie:
+				// RFC 8446, Section 4.2.2
+				if !readUint16LengthPrefixed(&extData, &m.cookie) ||
+					len(m.cookie) == 0 {
+					return false
+				}
+
+				// utls specific
+				// TODO: I believe this should never be sent with initial client hello
+				// end utls specific
+
+			case extensionKeyShare:
+				// RFC 8446, Section 4.2.8
+				var clientShares cryptobyte.String
+				if !extData.ReadUint16LengthPrefixed(&clientShares) {
+					return false
+				}
+				for !clientShares.Empty() {
+					var ks keyShare
+					if !clientShares.ReadUint16((*uint16)(&ks.group)) ||
+						!readUint16LengthPrefixed(&clientShares, &ks.data) ||
+						len(ks.data) == 0 {
+						return false
+					}
+					m.keyShares = append(m.keyShares, ks)
+				}
+
+				// utls specific
+				var ks keyShares = m.keyShares
+				clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &KeyShareExtension{ks.ToPublic()})
+				// end utls specific
+
+			case extensionEarlyData:
+				// RFC 8446, Section 4.2.10
+				m.earlyData = true
+
+				// utls specific
+				// TODO: read this into a generic extension
```

Comment:
I'm not sure we would want to carry this one through unless we had a fresh hello for each request.  ie it wouldn't be a fixed value between any two requests even in the case that the early data it was communicating was (which would also be questionable).

Reply:
Since deleting these is not length preserving it may indicate to do this perfectly we would also need to force a new hello for each request.

Reply:
Yeah so one other option is that we return an error if we find that the hello has one of these fields that we can't really copy and we can't properly imitate.